### PR TITLE
Classify lintish diagnostics as warnings, add treatWarningsAsErrors

### LINKED ISSUE #13408: Add support for diagnostic severities
TypeScript has added compiler options for lint level checks like `noUnusedParameters` or `noUnusedLocals` in tsconfig.json (which is goodness). Today when such an option is enabled then the corresponding issues are reported by TypeScript in the same way as semantic or syntax errors. There is no notion of severity. 

##  Background
We are using tslint to develop VS Code and we have used tslint rules to detect unused locals. TSLint rule failures are reported in VS code as warnings and we had the setup that lint style issue are shown as warnings and the typescript issues (syntax, semantic) are shown as errors. Now that typescript added more lint style checks we no longer get the distinction when developing between lint level warnings and semantic errors. The situation has become worse with tslint 4.0, there tslint has started to deprecate rules which are covered by TypeScript compiler options. This makes good sense, but it means we are now starting to see 'unused locals' reported as errors and no longer as warnings as we did before.

## Suggestion
Support that the user can define in the tsconfig.json whether a check enabled by compiler option should be reported as `error` or `warning`.

`eslint` supports to configure the severity of an option. Here is an example the [eslint documentation](http://eslint.org/docs/user-guide/configuring):

```json
{
    "rules": {
        "eqeqeq": "off",
        "curly": "warn",
        "quotes": ["error", "double"]
    }
}
```

// CC @waderyan @mjbvz

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #13408

We already had the infrastructure to do this in pretty much the entire stack (heck, `pretty` output even has configured yellow text for warnings, even though in no circumstances would we have seen them before). All that was missing was the ability to toggle warnings as errors and the initial listing of the diagnostics as warnings.

@mhegazy As a feature, I assume this is too late for 2.6?

### CODE REVIEW

File: tests/baselines/reference/treatWarningsAsErrorsFalse.errors.txt [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+              ~
+!!! warning TS6133: 'y' is declared but its value is never read.
+        switch (x.kind) {
+            case "a":
+            ~~~~
+!!! warning TS7029: Fallthrough case in switch.
+                void x; // implicit fallthrough
+            case "b":
+                 ~~~
+!!! error TS2678: Type '"b"' is not comparable to type '"a"'.
```
Comment (weswigham) [inline tests/baselines/reference/treatWarningsAsErrorsFalse.errors.txt]: This isn't strictly related to the test, but when I went to remove it, I considered that it is useful in verifying that we're not reporting _all_ errors as warnings.

File: src/compiler/program.ts [hunk 2]
```diff
@@ -1290,10 +1290,21 @@ namespace ts {
             });
         }
 
+        function shouldTreatWarningsAsErrors() {
+            const setValue = program.getCompilerOptions().treatWarningsAsErrors;
+            if (typeof setValue === "undefined") {
+                return true;
```
Comment (Kovensky) [inline src/compiler/program.ts]: Nice that it defaults to the safer setting (`true`); but how do you _disable_ it via command-line arguments? `--no-treatWarningsAsErrors`?

Comment (weswigham) [reply src/compiler/program.ts]: `--treatWarningsAsErrors false`, same as our other boolean flags.

**Comment by weswigham** [comment]
I _think_ I've gotten all the places where the warning v. error distinction matters changed. Also:
image
Note the green underline. Already works in vscode, too. We really did plumb this through almost _everywhere_ already.

**Comment by RyanCavanaugh** [comment]
What's the rule for whether something is a warning or not?

**Comment by weswigham** [comment]
> Based on the disccussion in #18894, a new flag --treatWraningsAsErrorrs will be added, defaults to true. The servirty of the style checks done by the compiler will all be set to Warrning. here is the list of configurations to control these messages :
--noUnusedLocals
--noUnusedParameters
--noImplicitReturns
--noFallthroughCasesInSwitch
--allowUnusedLabels
--allowUnreachableCode

For future errors... Based on this pattern, non-semantic errors which aren't runtime errors but are code smells.

**Comment by alexeagle** [comment]
I didn't check, but does the `tsc` output order the errors last? That is very important when a build has a wall of warnings, so I don't have to scroll around my terminal buffer or CI log.

**Comment by mhegazy** [comment]
> Also, I think it's desirable that the treatWarningsAsErrors applies to the exit code and on the command line, but not in the editor.

I am not sure i understand this statement. can you elaborate?

**Comment by weswigham** [comment]
@mhegazy I believe it means his disagrees with the changes [here](https://github.com/Microsoft/TypeScript/pull/19126/files#diff-54b7fc1f5e83aad83c6441769472ad4cR117), which make it so that with warnings `tsc` can still have a `0` exit code.

**Comment by weswigham** [comment]
> I didn't check, but does the tsc output order the errors last? That is very important when a build has a wall of warnings, so I don't have to scroll around my terminal buffer or CI log.

No, right now warnings are mixed with errors, sorted by file position. Is it always desirable to split them up?

Actually `rustc` only outputs warnings when there are no errors - is that more desirable? (Then again, `#deny` in rust also _only_ changes the exit code - warnings are still only reported once all errors are resolved... And in `rust` a program with errors is never compiler, with TS, even with errors you get output...)

**Comment by timc13** [comment]
any movement here?

**Comment by alexeagle** [comment]
@timc13 see the design notes linked above:
> Conclusion: hold off on dealing with warnings, understand how to improve TSLint, think about how to champion community tooling better by default for projects.

**Comment by RyanCavanaugh** [comment]
@weswigham can we close for now?

**Comment by weswigham** [comment]
We're not actively considering it right now, and I imagine the next time we decide to talk about this we'll waffle just as much, so sure. 😛 

The real question is will you close the original issue: #13408

**Pull request closed without merging**